cinn/hlir/op/custom_call.cc

```diff
@@ -0,0 +1,355 @@
+                                                  const Target &target) {
+  framework::CINNCompute compute([=](lang::Args args, lang::RetValue *ret) {
+    CHECK_EQ(args.size(), 1UL);
+    CINNValuePack pack_args = args[0];
+    CHECK(pack_args.back().is_string());
+
+    auto &attr_store = attrs.attr_store;
+    CHECK(attr_store.count("custom_call"));
+    std::string custom_call_api = absl::get<std::string>(attr_store.at("custom_call"));
+    auto args_func              = CustomCallArgsFuncRegistry::Global().Lookup(custom_call_api, target);
+
+    std::string func_name = pack_args.back().operator std::string();
+    // create call function.
+    ir::Var kernel_args(KERNEL_ARGS, type_of<void *>());
+    ir::Var kernel_args_num(KERNEL_ARGS_NUM, type_of<int>());
+
+    auto args_list                  = args_func(attrs, inputs, output_shapes);
+    std::vector<ir::Expr> host_args = {kernel_args, kernel_args_num};
+    host_args.insert(host_args.end(), args_list.begin(), args_list.end());
+    std::vector<ir::Argument> arguments = {ir::Argument(kernel_args, ir::Argument::IO::kOutput),
+                                           ir::Argument(kernel_args_num, ir::Argument::IO::kInput)};
+    // if target is nvgpu, add stream.
+    if (target == common::DefaultNVGPUTarget()) {
+      ir::Var kernel_stream(KERNEL_STREAM, type_of<void *>());
+
+      host_args.push_back(kernel_stream);
+      arguments.emplace_back(kernel_stream, ir::Argument::IO::kOutput);
+    }
+    auto call_extern_api =
+        ir::Call::Make(Void(), custom_call_api, host_args, {}, ir::CallType::Extern, ir::FunctionRef(), 0);
+    auto func = ir::_LoweredFunc_::Make(func_name, arguments, call_extern_api, {});
+
+    VLOG(3) << func;
+    *ret = CINNValuePack{{CINNValue(ir::Expr(func))}};
+  });
+
+  framework::CINNSchedule schedule([=](lang::Args args, lang::RetValue *ret) {});
+
+  auto strategy = std::make_shared<framework::OpStrategy>();
+  strategy->AddImpl(compute, schedule, "strategy.custom_call.x86", 1);
+  return strategy;
+}
+
+std::vector<ir::Expr> CustomCallArgsForCublas(const framework::NodeAttr &attrs,
+                                              const std::vector<ir::Tensor> &inputs,
+                                              const std::vector<std::vector<int>> &output_shapes) {
+  CHECK_EQ(inputs.size(), 2);
+  CHECK_EQ(output_shapes.size(), 1);
+
+  auto attr_store = attrs.attr_store;
+  bool trans_a    = attr_store.count("trans_a") ? absl::get<bool>(attr_store.at("trans_a")) : false;
+  bool trans_b    = attr_store.count("trans_b") ? absl::get<bool>(attr_store.at("trans_b")) : false;
+  float alpha     = attr_store.count("alpha") ? absl::get<float>(attr_store.at("alpha")) : 1.0f;
+  float beta      = attr_store.count("beta") ? absl::get<float>(attr_store.at("beta")) : 0.0f;
+  // b0 b1
+  int b0 = inputs[0]->shape.size() == 3 ? inputs[0]->shape[0].as_int32() : 1;
```

Comment:
matmul是支持多维shape输入的啊，感觉batch size应该这样计算：
```suggestion
  int b0 = inputs[0]->shape.size() >= 3 ? std::accumulate(inputs[0]->shape.begin(), inputs[0]->shape.end() - 2, 1, std::multiplies<int>()); : 1;
```

Reply:
下一个PR改吧